test/hotspot/jtreg/serviceability/HeapDump/DuplicateArrayClassesTest.java

```diff
@@ -0,0 +1,334 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.io.BufferedInputStream;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Hashtable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+import jdk.test.lib.Asserts;
+import jdk.test.lib.JDKToolLauncher;
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.hprof.parser.PositionDataInputStream;
+import jdk.test.lib.process.ProcessTools;
+
+/**
+ * @test
+ * @bug 8281267
+ * @summary Verifies heap dump does not contain duplicate array classes
+ * @library /test/lib
+ * @run driver DuplicateArrayClassesTest
+ */
+
+class DuplicateArrayClassesTarg extends LingeredApp {
+    public static void main(String[] args) {
+        // Initialize some array classes (for primitive type and object type).
+        int[][][] intArray = new int[0][][];
+        String[][][] strArray = new String[0][][];
+        LingeredApp.main(args);
+        System.out.println("" + intArray + strArray);   // to be sure the classes are not unloaded
+    }
+}
+
+
+public class DuplicateArrayClassesTest {
+
+    public static void main(String[] args) throws Exception {
+        File dumpFile = new File("Myheapdump.hprof");
+        createDump(dumpFile);
+        verifyDump(dumpFile);
+    }
+
+    private static void createDump(File dumpFile) throws Exception {
+        LingeredApp theApp = null;
+        try {
+            theApp = new DuplicateArrayClassesTarg();
+
+            LingeredApp.startApp(theApp);
+
+            //jcmd <pid> GC.heap_dump <file_path>
+            JDKToolLauncher launcher = JDKToolLauncher
+                    .createUsingTestJDK("jcmd")
+                    .addToolArg(Long.toString(theApp.getPid()))
+                    .addToolArg("GC.heap_dump")
+                    .addToolArg(dumpFile.getAbsolutePath());
+            Process p = ProcessTools.startProcess("jcmd", new ProcessBuilder(launcher.getCommand()));
+            // If something goes wrong with heap dumping most likely we'll get crash of the target VM.
+            while (!p.waitFor(5, TimeUnit.SECONDS)) {
```

Comment:
5 seconds seems kind of short. Doesn't allow for occasional network hiccups.

Reply:
Note that this is a cycle. It ensures target VM (theApp) is not terminated while jcmd (p) processes the command.
I had some experience in the past when jcmd didn't exit if target VM crashes.